### PR TITLE
New Goodie IA (Lorem Ipsum Generator)

### PR DESCRIPTION
**Pull request by jee1mr**
Continuation of #1005

### CODE REVIEW

**Comment by moollaza** [comment]
@jee1mr this is great. We can make this even more awesome though, because we now have templates for Goodies!

I took a minute to convert this to a templated Goodie and pushed to a branch for you. Please checkout the diff at https://github.com/duckduckgo/zeroclickinfo-goodies/commit/67befb0ea046c69385f0fdec58397068cd07a298?diff=split

If you update the .pm accordingly and create a new sharedir, `/share/lorem_ipsum` containing the template `content.handlebars`

You'll be using the new, preferred, structured answer. It'll look like this:

2015-04-16_09h40_27

------

I didn't update the tests, but you can easily do that. Here's a great example: https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/t/Minecraft.t

@abeyang, this is ready for a design signoff.

**Comment by jee1mr** [comment]
@moollaza Thanks for the help.
Now how do I go about with this? Should I merge your commit with my repo? Or should I make the necessary changes manually?
I'll update the tests.

**Comment by moollaza** [comment]
Probably simplest to make the changes manually, I'm not sure if you can cherry-pick that commit...

**Comment by jee1mr** [comment]
@moollaza Thanks for the help.
Now how do I go about with this? Should I merge your commit with my repo? Or should I make the necessary changes manually?
I'll update the tests.

**Comment by jee1mr** [comment]
@moollaza Damn! Why the merge conflict again?

**Comment by moollaza** [comment]
@jee1mr we just merged a pr today that updated the dist.ini -- you no longer need to add Perl packages to the dist.ini so likely your commit and the update have conflicted. You should be able to merge master into your branch and simply remove the changes you made to the dist.ini

**Comment by jee1mr** [comment]
@moollaza Done :+1: Thanks :)
I'll update the tests soon.

File: dist.ini [hunk 1]
```diff
@@ -13,7 +13,105 @@ index = http://duckpan.org
 module = Dist::Zilla::Plugin::UploadToDuckPAN
 module = Dist::Zilla::Plugin::BuildShareAssets
 
+<<<<<<< HEAD
```
Comment (moollaza) [inline dist.ini]: Whoops, looks like you committed the dist with the merge conflicts in there.

You'll want to remove everything from `<<<<<<HEAD` to `==========`

We're replacing the `[Prereqs]` and `[Prereqs / TestRequires]` with `[AutoPrereqs]`

**Comment by moollaza** [comment]
@jee1mr the Perl dep has been added to the backend, once the merge conflict is removed this should be good to move ahead :+1:

File: dist.ini [hunk 2]
```diff
@@ -14,6 +14,7 @@ module = Dist::Zilla::Plugin::UploadToDuckPAN
 module = Dist::Zilla::Plugin::BuildShareAssets
 
 [AutoPrereqs]
+>>>>>>> a3a9ccfecc4c25b88c3979294c6733be2df2658d
```
Comment (mintsoft) [inline dist.ini]: This `>>>>>>> a3a9ccfecc4c25b88c3979294c6733be2df2658d` is an artefact from the merge, it should be removed

**Comment by moollaza** [comment]
LGTM :+1:

**Comment by moollaza** [comment]
Update, it was already added!

**Comment by moollaza** [comment]
Actually, this still needs a design signoff!

/cc @abeyang

**Comment by abeyang** [comment]
Nice @jee1mr, we're almost there! Just a few tweaks:

1. Is it possible to add a couple more sentences to each paragraph? Some of them are pretty small... 
2. By default, let's just return **3 paragraphs**
3. Also, when doing default search, can we add a pipe and the phrase, "Try lorem ipsum 8 for more"? (@moollaza is this doable?) This should only show during default search, and hidden when they've searched with another number. Screenshot below.
4. When it's just one paragraph, it should say "1 Random Paragraph" (right now it's showing plural, "paragraphs")

screen shot 2015-04-28 at 4 40 30 pm

Thanks! Let us know if you need any help!

**Comment by moollaza** [comment]
>Also, when doing default search, can we add a pipe and the phrase, "Try lorem ipsum 8 for more"? (@moollaza is this doable?)

Yup, its doable if we switch to using a template for the subtitle:

```perl
        templates => {
            group => 'text',
            options =>{
                subtitle_content => 'DDH.lorem_ipsum.subtitle',
                content => 'DDH.lorem_ipsum.content',
                moreAt => 1
            }
        }
```

**Comment by jee1mr** [comment]
@moollaza Struggling with the test file. Need help.

**Comment by jee1mr** [comment]
@abeyang 
default

singular

**Comment by jee1mr** [comment]
@abeyang Also increased paragraph size. Is this fine or need more?

**Increased Paragraph size**

size

**Comment by abeyang** [comment]
I think these changes, including the paragraph size, is good! :+1: /cc @moollaza

**Comment by jee1mr** [comment]
@abeyang Also increased paragraph size. Is this fine?

**Increased Paragraph size**

size

**Comment by abeyang** [comment]
Oh did you increase it again? This one seemed right to me:

screen shot 2015-04-29 at 9 36 36 am

**Comment by jee1mr** [comment]
Done.

**Comment by moollaza** [comment]
:+1:

File: t/LoremIpsum.t [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+use DDG::Test::Goodie;
+
+zci answer_type => "lorem_ipsum";
+zci is_cached   => 1;
+
+ddg_goodie_test(
+    [qw( DDG::Goodie::LoremIpsum )],
+    'lorem ipsssum 3' => undef,
+    'lorem dipsum' => undef,
+    'lipsum 10' => test_zci(
```
Comment (moollaza) [inline t/LoremIpsum.t]: Would suggest you also have a test for "lorem ipsum" to verify the minimum number of paras works. In general we like to see tests for the example primary queries at least.

**Comment by jee1mr** [comment]
@moollaza Yeah. I'll add all possible cases. But I wasn't able to figure out how to match the array with regex.

File: t/LoremIpsum.t (hunk 3 above)
Comment (mintsoft) [inline t/LoremIpsum.t]: I'd probably be a little stronger and say "need tests for the example primary queries" !

**Comment by jee1mr** [comment]
@mintsoft @moollaza Yeah, I was aware of that :smiley: Check previous commits on test file.
But after we switched to using template, tests started failing. So, I wanted to first work on rectifying the test file and then move on. That's why there's only one test case.

**Comment by moollaza** [comment]
@jee1mr looks like tests are passing now?

**Comment by jee1mr** [comment]
@moollaza No.

```
[02:02 PM codio@ninja-bottle zeroclickinfo-goodies {master}]$ prove -Ilib t/LoremIpsum.t                                                     
t/LoremIpsum.t .. 1/?                                                                                                                        
    #   Failed test 'Deep: full ZCI object'                                                                                                  
    #   at /home/codio/perl5/perls/perl-5.18.2/lib/site_perl/5.18.2/DDG/Test/Goodie.pm line 76.                                              
    #     Structures begin differing at:                                                                                                     
    #          $got->{structured_answer}{data}{lorem_array}[0] = 'Nobis tempore consequatur tempore recusandae voluptate placeat illum. Itaqu
e neque non quia nam qui eaque libero et. Non porro omnis id.Laboriosam adipisci ex vitae ut perspiciatis excepturi quod cupiditate. Adipisci
 dolor sapiente laborum voluptatibus aspernatur nam. Libero sit omnis delectus. Sunt id sit labore fuga sunt unde. Ut a aperiam nisi amet sin
t architecto placeat. Ut similique a est occaecati voluptatibus ad sed.'                                                                     
    #     $expected->{structured_answer}{data}{lorem_array}[0] = (?^:[a-zA-Z \.\n]*)                                                         
    # Looks like you failed 1 test of 2.                                                                                                     
                                                                                                                                             
#   Failed test 'Query: lipsum 10'                                                                                                           
#   at /home/codio/perl5/perls/perl-5.18.2/lib/site_perl/5.18.2/DDG/Test/Goodie.pm line 80.                                                  
# Looks like you failed 1 test of 3.                                                                                                         
t/LoremIpsum.t .. Dubious, test returned 1 (wstat 256, 0x100)                                                                                
Failed 1/3 subtests                                                                                                                          
                                                                                                                                             
Test Summary Report                                                                                                                          
-------------------                                                                                                                          
t/LoremIpsum.t (Wstat: 256 Tests: 3 Failed: 1)                                                                                               
  Failed test:  3                                                                                                                            
  Non-zero exit status: 1                                                                                                                    
Files=1, Tests=3,  0 wallclock secs ( 0.02 usr  0.01 sys +  0.19 cusr  0.03 csys =  0.25 CPU)                                                
Result: FAIL                      
```

**Comment by moollaza** [comment]
@jee1mr it seems that testing the content of the array isn't really going to be possible with our current testing framework. 

For now, the following will suffice:

```perl
#!/usr/bin/env perl

use strict;
use warnings;
use Test::More;
use Test::Deep;
use DDG::Test::Goodie;

zci answer_type => "lorem_ipsum";
zci is_cached   => 1;

ddg_goodie_test(
    [qw( DDG::Goodie::LoremIpsum )],
    'lorem ipsssum 3' => undef,
    'lorem dipsum' => undef,
    'lipsum 10' => test_zci(
        qr/[a-zA-Z .]+/,
        make_structured_answer(2)
    )

);

sub make_structured_answer {
    my $loop = $_;

    return structured_answer => {
        id => 'lorem_ipsum',
        name => 'Answer',
        data => "-ANY-"
         meta => {
            sourceName => "Lipsum",
            sourceUrl => "http://lipsum.com/"
        },
        templates => {
            group => 'text',
            options => {
                subtitle_content => 'DDH.lorem_ipsum.subtitle',
                content => 'DDH.lorem_ipsum.content',
                moreAt => 1
            }
        }
    };
};

done_testing;
```

---

I'll be discussing the testing framework with the team so that we can improve it to handle these newer, complex `structured_answer` hashes.

**Comment by jee1mr** [comment]
@moollaza Done. :+1:

**Comment by moollaza** [comment]
LGTM :+1:

**Comment by moollaza** [comment]
@jee1mr I can't believe it, but unfortunately while doing some further testing of this IA in our production environment I ran into issue and I'm so far, unable to get this IA working.

I wrestled with it for a few hours yesterday, and had no luck. For some reason, when the Text::Lorem package is generating the paragraphs of text, its running into an `uninitialized index` error, meaning it's trying to access undefined array elements. I'm not sure how or why this is happening. I looked at the Text::Lorem code, which is pretty simple and didn't notice anything. I even tried switching to the Test::Lorem::More package and had no luck.

I tried fully re-writing the IA and still the package was throwing these errors, so for now I can't deploy it live. I'm terribly sorry this has happened! I'll run it by the engineering team to see if they can figure it out, but it may be best to switch this IA to a Spice and use an API.

I'll get back to you in the week if I have any progress diagnosing the issue!

**Comment by jee1mr** [comment]
@moollaza Sorry. I was busy with my semester exams.
If nobody else has taken this up yet, I'll convert this to Spice during the DDG Meetup(B'lore) this Sunday.

**Comment by jee1mr** [comment]
@moollaza But isn't there a spice already internally? It needed to be converted to a Goodie and made open. That's how I started working on it.

**Comment by moollaza** [comment]
@jee1mr so sorry! I've fallen behind and just seen this!

Right now it's a hardcoded, internal "answer" that was made several years ago. It can be converted to either a Spice or Goodie. It's clear now that the Goodie route isn't going to work as well as we'd hoped though!

A Spice using an API would be fine. We already have a "baconipsum" Spice, but we still need a proper "loremipsum' Spice, that uses a JSON API.
